rfcs/0012-compaction-state-persistence.md

```diff
@@ -0,0 +1,471 @@
+- **DELETE operations**: $0.0005 per 1,000 requests
+
+**Baseline costs:**
+- **160 PUTs** (SST writes): $0.000080
+- **160 GETs** (SST reads): $0.000064
+- **2 PUTs** (manifest): $0.000001
+- **Total baseline**: $0.000145
+
+**Additional persistence costs:**
+- **160 PUTs** (state writes): $0.000080
+- **Additional total**: $0.000080
+
+**Cost impact:**
+- **Additional cost**: $0.000080 (~$0.00008)
+- **Percentage increase**: +50% operations, but negligible absolute cost
+- **Cost per GB**: ~$0.000002 per GB compacted
+
+
+### **Recovery Efficiency Analysis**
+
+#### **Work Preservation Calculation**
+
+**Without persistence:**
+- **Recovery strategy**: Restart compaction from beginning
+- **Work preserved**: 0% (all progress lost)
+- **Additional operations**: Full re-execution (322 operations repeated)
+- **Time impact**: 100% of original compaction time
+
+**With persistence:**
+- **Average failure point**: 50% through compaction (statistical)
+- **Work preserved**: ~50% of progress maintained
+- **Recovery operations**: Resume from last checkpoint
+- **Time impact**: ~50% of original compaction time saved
+
+**Detailed recovery scenarios:**
+
+| Failure Point | Without Persistence | With Persistence | Work Preserved | Operations Saved(work preserved percentage * 322) |
+|---------------|-------------------|------------------|----------------|------------------|
+| **10% complete** | Restart (0% preserved) | Resume from 10% (10% preserved) | 10% | 32 operations |
+| **25% complete** | Restart (0% preserved) | Resume from 25% (25% preserved) | 25% | 81 operations |
+| **50% complete** | Restart (0% preserved) | Resume from 50% (50% preserved) | 50% | 161 operations |
+| **75% complete** | Restart (0% preserved) | Resume from 75% (75% preserved) | 75% | 242 operations |
+| **90% complete** | Restart (0% preserved) | Resume from 90% (90% preserved) | 90% | 290 operations |
+
+**Average work preservation**: **50%** across all failure scenarios
+
+#### **Scaling Analysis**
+
+| Compaction Size | SSTs | Base Ops | +Persistence | Additional Cost | % Overhead |
+|-----------------|------|----------|--------------|----------------|------------|
+| **10GB** (40 SSTs) | 40 | 82 | 122 | $0.000020 | +49% |
+| **40GB** (160 SSTs) | 160 | 322 | 482 | $0.000080 | +50% |  
+| **100GB** (400 SSTs) | 400 | 802 | 1,202 | $0.000200 | +50% |
+| **1TB** (4,000 SSTs) | 4,000 | 8,002 | 12,002 | $0.002000 | +50% |
+
+**Key observations:**
+- **Operations increase by ~50%**, but absolute costs remain minimal
+- **Cost scales linearly** with compaction size (~$0.000002/GB)
+- **Percentage overhead is consistent** at ~50% across all sizes  
+- **Total costs are negligible** compared to storage and compute costs
+
+
+## Future Extensions
+
+### **Distributed Compaction**
+- Persistent state provides foundation for multi-compactor coordination and work distribution.
+- Add time boundary based persistence of SSTs to the compaction state in the object_store
```

Comment:
```suggestion
- Define a minimum time boundary between compaction file updates to prevent excessive writes to the file (see https://github.com/slatedb/slatedb/pull/695#discussion_r2229977189)
```